hands-on-exercise/1-native/2-build-module.md

```diff
@@ -0,0 +1,947 @@
+```
+
+You can now define the module:
+
+```sh
+$ cd checkers-minimal
+$ touch go.mod
+```
+
+In the new file, you define your module name and the Go version:
+
+```txt
+module github.com/alice/checkers
+
+go 1.21
+```
+
+## Protobuf files
+
+On top of the _regular_ Protobuf files, like message and query types, your checkers module type will also be identified with a Protobuf file. You have chosen the package `github.com/alice/checkers`, so go ahead and create the necessary folders:
+
+```sh
+$ mkdir -p proto/alice/checkers/v1
+$ mkdir -p proto/alice/checkers/module/v1
+```
+
+### Module proto
+
+You define your module in a new `proto/alice/checkers/module/v1/module.proto`, taking inspiration from `minimal-module-example`:
+
+```protobuf [proto/alice/checkers/module/v1/module.proto]
+syntax = "proto3";
+
+package alice.checkers.module.v1;
+
+import "cosmos/app/v1alpha1/module.proto";
+
+// Module is the app config object of the module.
+// Learn more: https://docs.cosmos.network/main/building-modules/depinject
+message Module {
+  option (cosmos.app.v1alpha1.module) = {
+    go_import : "github.com/alice/checkers"
+  };
+
+  // authority defines the custom module authority.
+  // if not set, defaults to the governance module.
+  string authority = 1;
+}
+```
+
+Note how the package is `alice.checkers.module.v1` and the message name is `Module`. This means that later on, when you integrate the module in your chain, you will identify it with:
+
+```txt
+alice.checkers.module.v1.Module
+```
+
+### Necessary Protobuf files
+
+The compilation of your Protobuf files will be done with the `scripts/protocgen.sh` script, which you can copy from `minimal-module-example`:
+
+```sh [scripts/protocgen.sh]
+#!/usr/bin/env bash
+
+set -e
+
+echo "Generating gogo proto code"
+cd proto
+proto_dirs=$(find . -path -prune -o -name '*.proto' -print0 | xargs -0 -n1 dirname | sort | uniq)
+for dir in $proto_dirs; do
+  for file in $(find "${dir}" -maxdepth 1 -name '*.proto'); do
+    # this regex checks if a proto file has its go_package set to github.com/alice/checkers/api/...
+    # gogo proto files SHOULD ONLY be generated if this is false
+    # we don't want gogo proto to run for proto files which are natively built for google.golang.org/protobuf
+    if grep -q "option go_package" "$file" && grep -H -o -c 'option go_package.*github.com/alice/checkers/api' "$file" | grep -q ':0$'; then
+      buf generate --template buf.gen.gogo.yaml $file
+    fi
+  done
+done
+
+echo "Generating pulsar proto code"
+buf generate --template buf.gen.pulsar.yaml
+
+cd ..
+
+cp -r github.com/alice/checkers/* ./
+rm -rf api && mkdir api
+mv alice/checkers/* ./api
+rm -rf github.com alice
+```
+
+Note how it mentions two files `buf.gen.gogo.yaml` and `buf.gen.pulsar.yaml`. So go ahead and copy them from `minimal-module-example`. In fact, also copy `buf.yaml` and the version lock file to avoid surprises:
```

Comment:
@xavierlepretre Given that we now the relative location of the folders, perhaps you can include the tree `cp` commands directly into the content for an easy copy&paste.

Reply:
Good point. Will do.